src/xdp-utils.c

```diff
@@ -144,11 +145,72 @@ xdp_app_info_new (XdpAppInfoKind kind)
   return app_info;
 }
 
+void
+decode_hex_escapes (char *str[])
+{
+  char hex[] = "0xhh";
+
+  for (int i=0; (*str)[i]!='\0' && (*str)[i+1]!='\0'; ++i)
+  {
+    if ((*str)[i] == '\\' && (*str)[i+1] == 'x')
+      {
+        hex[2] = (*str)[i+2];
+        hex[3] = (*str)[i+3];
+        char unescaped = (char) strtol (hex, NULL, 16);
+        (*str)[i+2] = '\\';
+        (*str)[i+3] = unescaped;
+        i += 3;
+      }
+  }
+  char **splits = g_strsplit(*str, "\\x\\", -1);
+  g_free (*str);
+  *str = g_strjoinv ("", splits);
+  g_strfreev (splits);
+  return;
+}
+
+void
+get_app_id_from_pid (pid_t pid, XdpAppInfo *app_info)
+{
+  char *unit, **dash_splits, **at_splits, **app_id;
+  int res, len;
+
+  res = sd_pid_get_user_unit(pid, &unit);
+  /*
+   * the session might not be managed by systemd
+   * or there could be error fetching own systemd units
+   * or the unit might not be started by the the desktop environment (e.g. it's a script run from terminal)
+   */
+  if (res == -ENODATA || res < 0 || !g_str_has_prefix (unit, "app-"))
+    {
+      app_info->id = g_strdup ("");
+      return;
+    }
+  dash_splits = g_strsplit (unit, "-", -1);
+  len = g_strv_length (dash_splits);
+  /*
+   * format:
+   * app[-<launcher>]-<ApplicationID>-<RANDOM>.scope
+   * app[-<launcher>]-<ApplicationID>-autostart.service -> no longer true since systemd v248
+   */
+  if (g_str_has_suffix (unit, ".scope") || g_str_has_suffix (unit, "-autostart.service"))
+    app_id = &dash_splits[len-2];
+  /* format: app[-<launcher>]-<ApplicationID>[@<RANDOM>].service */
+  else {
+    at_splits = g_strsplit (dash_splits[len-1], "@", 2);
+    app_id = &at_splits[0];
+  }
+  decode_hex_escapes (app_id);
+  app_info->id = g_strdup (*app_id);
+  g_strfreev(at_splits);
+  g_strfreev (dash_splits);
```

Comment:
You can remove these calls by declaring these variables as `g_auto(GStrv) at_splits = NULL` and `g_auto(GStrv) dash_splits = NULL`